### PR TITLE
[download] Migrate to Pulsar's Rolling Release Repo, off of CirrusCI

### PR DESCRIPTION
**Pull request by confused-Techie**
This PR removes our `download` microservices interaction with CirrusCI, instead now collecting rolling releases from [`pulsar-edit/pulsar-rolling-releases`](https://github.com/pulsar-edit/pulsar-rolling-releases) repository.

These changes relate to the necessary changes needed to be made, detailed in [`pulsar-edit/pulsar#685`](https://github.com/pulsar-edit/pulsar/issues/685).

And should resolve any and all issues related to the download microservice, to ensure uninterrupted functionality of our download microservice.

### CODE REVIEW

**Comment by DeeDeeG** [comment]
In my ideal design in my head, the download microservice would branch its logic based on "ARM Linux/Apple Silicon macOS? --> get from Cirrus", "other stuff? --> get from GitHub Releases rolling".

So that the download microservice _still works_ for resolving Apple Silicon and ARM Linux payload URLs. Otherwise the service is dropping our ARM builds.

Haven't taken a close look, but it's nice if this is working already.

**Comment by confused-Techie** [comment]
@DeeDeeG You do have a really good point there.

Although what if we take a slightly different approach and instead add scripts to the end of builds on cirrus to upload those to the github releases? Since you can see in the PR the GitHub API is much simpler to work with, and would be nice to have a proper visual archive of all our rolling releases.

Although I do get if that's difficult to do and we have to add back in the capabilities to work with Cirrus. Since that wasn't something fully considered during the drafting of this PR. Sure this PR will still look for ARM and Silicon binaries, but I had forgotten that we had no plans to make them available here

**Comment by DeeDeeG** [comment]
:shrug: It's a little bit down to what we can _dream up_, _design_ and _implement_ in the next few days. Not trivial, theoretically doable. I'm sure we can add back more stuff if some of it gets left on the cutting room floor in order to get something out the door in time. Something is better than nothing, to be sure.

**Comment by confused-Techie** [comment]
Appreciate you! But very true, we are limited by time.

Although over on the GitHub Actions PR on the main repo, I put all the logic of uploading artifacts into it's own NodeJS script, not relying on much else, so it may be totally possible to get Cirrus to run that same script after the fact, I'll actually take a try at implementing that soon, since if so it'd be rad to get this all functioning without to many more logic changes lol

File: microservices/download/utils.js [hunk 1]
```diff
@@ -111,175 +107,193 @@ async function displayError(req, res, errMsg) {
...
-              }
-            }
-          }
+
+    let releases = await doRequest();
+
+    // Now these releases should be sorted already, if we find they aren't we might
+    // have to add semver as a dep on this microservice, which is no fun since this
+    // microservice has 0 deps currently. For now lets assume it's a sorted array
+    // This same assumption is made on the `pulsar-updater` core package
```
Comment (DeeDeeG) [inline microservices/download/utils.js:116]: NO ACTION NEEDED JUST REMARKING:

Sounds reasonable to me, and the releases do seem to be in chronological order (newest-first) in the JSON response. 👍 LGTM.

File: microservices/download/utils.js [hunk 2]
```diff
@@ -111,175 +107,193 @@ async function displayError(req, res, errMsg) {
...
+    // So we will return an error
 
-    // Now that we have the binary, it's time to return a redirect.
     return {
-      ok: true,
-      content: `https://api.cirrus-ci.com/v1/artifact/task/${taskid}/binary/${binaryPath}`
+      ok: false,
+      code: 404,
+      msg: `Unable to find any assets matching the provided parameters: os=${os};type=${type}`
     };
```
Comment (DeeDeeG) [inline microservices/download/utils.js:297]: NO ACTION NEEDED, JUST COMMENTING 👍 

This "asset not found" handling _does_ work in my testing. Some of the blockmap asset types aren't available, like the intel macOS and the Windows ones, and if you request one of those, the microservice legitimately can't find them, and the error message is returned as it should be in that situation. 👍 

Not the microservice's fault those assets genuinely aren't available, so 404'ing is the right move. 👍  And it does work as such in my testing. 👍

File: microservices/download/utils.js [hunk 3]
```diff
@@ -111,175 +107,193 @@ async function displayError(req, res, errMsg) {
...
+
+    // Now these releases should be sorted already, if we find they aren't we might
+    // have to add semver as a dep on this microservice, which is no fun since this
+    // microservice has 0 deps currently. For now lets assume it's a sorted array
+    // This same assumption is made on the `pulsar-updater` core package
+
+    for (let version of releases) {
+      for (let asset of version.assets) {
+
+        let name = asset?.name;
```
Comment (DeeDeeG) [inline microservices/download/utils.js:121]: MINOR CODE NITPICK

Use of optional chaining here isn't really productive with how optional chaining works.

<details>

Optional chaining here checks that trying to access `asset.name` doesn't error of `asset` is undefined, but doesn't ensure that `asset.name` isn't undefined.

So, optional chaining isn't really applicable here, the body of this loop will only execute when iterating over `asset`s that do exist.

</details>

Can remove the optional chaining if you want, it's not hurting anything but it's not doing anything either I think.

```suggestion
        let name = asset.name;
```

And just have the loop handle if `name` turns out to be `undefined`, as you already do just below, that's a good approach already, IMO.

File: microservices/download/utils.js [hunk 4]
```diff
@@ -111,175 +107,193 @@ async function displayError(req, res, errMsg) {
...
+
+        let returnObj = {
+          ok: true,
+          content: asset?.browser_download_url
+        };
+
+        // Ensure we have valid data to work with
+        if (typeof name !== "string" && typeof returnObj.content !== "string") {
+          continue;
         }
```
Comment (DeeDeeG) [inline microservices/download/utils.js:131]: Is this meant to be an OR `||` check? Currently an AND `&&`.

<details>

If no `name` OR no `returnObj.content`? Or is "both being missing" the real indicator of bad data?

Is a binary (an `asset`) with _just_ a `name` or _just_ a `browser_download_url` possible/valid? Should lacking _one_ of those cause the loop to `continue` to try the next asset?

</details>

```suggestion
        // Ensure we have valid data to work with
        if (typeof name !== "string" || typeof returnObj.content !== "string") {
          continue;
        }
```

^ I'm assuming this is what is intended? Bail if _either_ `name` or `browser_download_url` is missing, and try the next asset instead?

**Review by DeeDeeG — COMMENTED** [review]
Some stuff I noticed, most of it is either really minor or just optional nitpicks.

There's one `&&` check that I think is meant to be `||`, but other than that, I think this looks pretty good! Worked in my testing already, as-is!

Thank you for this!!!!

File: microservices/download/utils.js (hunk 2 above)
Comment (confused-Techie) [reply microservices/download/utils.js:297]: Part of the reason I originally didn't upload them, made the 404 easy to test, so happy you were able to confirm

**Comment by DeeDeeG** [comment]
Posting my test results, as always for this microservice's PR's:

My old `test-find-link.js` from before (see code snippet below) totally works with the new download microservice code. So, this seems to be a functional drop-in replacement for the old code. (Just using GitHub Releases instead of Cirrus's API.)

```js
// This is test-find-link.js
const utils = require("./utils.js");

(async () => {
  console.log(await utils.findLink("arm_linux", "linux_deb"));
})();
```

<details><summary>Test output when modifying and running <code>node ./test-find-link</code> to test all possible binary assets (click to expand):</summary>

```console
% time node ./test-find-link.js 
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082516/Pulsar-1.108.2023082516.AppImage'
}
node ./test-find-link.js  0.12s user 0.04s system 37% cpu 0.410 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082516/pulsar-1.108.2023082516.tar.gz'
}
node ./test-find-link.js  0.12s user 0.03s system 54% cpu 0.271 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082516/pulsar-1.108.2023082516.x86_64.rpm'
}
node ./test-find-link.js  0.12s user 0.03s system 56% cpu 0.269 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082516/pulsar_1.108.2023082516_amd64.deb'
}
node ./test-find-link.js  0.12s user 0.04s system 37% cpu 0.413 total
% time node ./test-find-link.js
{
  ok: false,
  code: 404,
  msg: 'Unable to find any assets matching the provided parameters: os=linux;type=windows_setup'
}
node ./test-find-link.js  0.12s user 0.03s system 53% cpu 0.274 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082517/Pulsar.Setup.1.108.2023082517.exe'
}
node ./test-find-link.js  0.12s user 0.03s system 35% cpu 0.406 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082517/Pulsar-1.108.2023082517-win.zip'
}
node ./test-find-link.js  0.12s user 0.03s system 56% cpu 0.264 total
% time node ./test-find-link.js
{
  ok: false,
  code: 404,
  msg: 'Unable to find any assets matching the provided parameters: os=windows;type=windows_blockmap'
}
node ./test-find-link.js  0.12s user 0.03s system 59% cpu 0.250 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082516/Pulsar-1.108.2023082516-mac.zip'
}
node ./test-find-link.js  0.12s user 0.03s system 38% cpu 0.378 total
% time node ./test-find-link.js
{
  ok: false,
  code: 404,
  msg: 'Unable to find any assets matching the provided parameters: os=intel_mac;type=mac_zip_blockmap'
}
node ./test-find-link.js  0.12s user 0.04s system 56% cpu 0.277 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082516/Pulsar-1.108.2023082516.dmg'
}
node ./test-find-link.js  0.12s user 0.03s system 53% cpu 0.270 total
% time node ./test-find-link.js
{
  ok: false,
  code: 404,
  msg: 'Unable to find any assets matching the provided parameters: os=intel_mac;type=mac_dmg_blockmap'
}
node ./test-find-link.js  0.12s user 0.03s system 52% cpu 0.275 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/Pulsar-1.108.2023082808-arm64-mac.zip'
}
node ./test-find-link.js  0.12s user 0.04s system 37% cpu 0.416 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/Pulsar-1.108.2023082808-arm64-mac.zip.blockmap'
}
node ./test-find-link.js  0.11s user 0.03s system 58% cpu 0.246 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/Pulsar-1.108.2023082808-arm64.dmg'
}
node ./test-find-link.js  0.12s user 0.03s system 37% cpu 0.389 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/Pulsar-1.108.2023082808-arm64.dmg.blockmap'
}
node ./test-find-link.js  0.12s user 0.03s system 55% cpu 0.271 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/Pulsar-1.108.2023082808-arm64.AppImage'
}
node ./test-find-link.js  0.12s user 0.03s system 35% cpu 0.433 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/pulsar-1.108.2023082808-arm64.tar.gz'
}
node ./test-find-link.js  0.12s user 0.03s system 59% cpu 0.260 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/pulsar-1.108.2023082808.aarch64.rpm'
}
node ./test-find-link.js  0.11s user 0.03s system 34% cpu 0.414 total
% time node ./test-find-link.js
{
  ok: true,
  content: 'https://github.com/pulsar-edit/pulsar-rolling-releases/releases/download/v1.108.2023082808/pulsar_1.108.2023082808_arm64.deb'
}
node ./test-find-link.js  0.12s user 0.03s system 39% cpu 0.379 total

```

</details>

Note that there are some `ok: false` results where I gave incorrect parameter combinations, or where there was a genuine and legitimate reason to return a 404, which worked as intended and as written, even without doing anything with the stuff I mentioned in my review comments.

**In summary:** I didn't find any real-world failures of the microservice logic with this testing. Handling bad input parameters and handling missing assets were also good and both working as intended in these tests.

File: microservices/download/utils.js (hunk 4 above)
Comment (confused-Techie) [reply microservices/download/utils.js:131]: Fantastic catch! You are totally correct, we need to bail if either of these is invalid, glad you saw this!

File: microservices/download/utils.js (hunk 3 above)
Comment (confused-Techie) [reply microservices/download/utils.js:121]: You do have a valid point here, without the use of optional chaining, if this item doesn't exist then we just get `undefined` and that is caught during a later check, as we wouldn't ever have an `asset` that itself is undefined.

So you correct this can be removed, although I'll say just for me personally adding things like this helps with not trusting the data we have received to be what it says it is, enforcing the unknown factor of this variable. 

But from a quality standpoint, it's pointless, and could potentially add confusion on how optional chaining operators work, so I'll side with you and remove it

File: microservices/download/utils.js [hunk 5]
```diff
@@ -111,175 +107,193 @@ async function displayError(req, res, errMsg) {
...
-          }
+
+    let releases = await doRequest();
+
+    // Now these releases should be sorted already, if we find they aren't we might
+    // have to add semver as a dep on this microservice, which is no fun since this
+    // microservice has 0 deps currently. For now lets assume it's a sorted array
+    // This same assumption is made on the `pulsar-updater` core package
+
+    for (let version of releases) {
```
[the remark this answers is not in the document]
Comment (confused-Techie) [reply microservices/download/utils.js:118]: Fair enough, I'll go ahead and define this as `const`. This one was simply defined without thinking the rest of the code through

**Comment by confused-Techie** [comment]
@DeeDeeG Addressed everything you commented on, if you'd like to do one more review so we can push this. Apparently our changes to cirrus have in fact already broken the download microservice

File: microservices/download/utils.js (hunk 3 above)
Comment (DeeDeeG) [reply microservices/download/utils.js]: I figured out there is one case, if the iterable (usually an array) contains explicit `undefined`, then the optional chaining _would_ be useful.

It happens to be safe not to do the optional chaining since we're getting this from JSON, and `undefined` isn't allowed in JSON.

But your initial instinct was probably about right, my apologies. I think I was a little off-base on this one posting without being expert in it, so if you wish to add it back that is reasonable. (Since we're getting this input data from JSON, it should be okay without, in this case.)

File: microservices/download/utils.js [hunk 6]
```diff
@@ -111,175 +107,193 @@ async function displayError(req, res, errMsg) {
...
+
+    // Now these releases should be sorted already, if we find they aren't we might
+    // have to add semver as a dep on this microservice, which is no fun since this
+    // microservice has 0 deps currently. For now lets assume it's a sorted array
+    // This same assumption is made on the `pulsar-updater` core package
+
+    for (const version of releases) {
+      for (const asset of version.assets) {
+
+        let name = asset.name;
```
Comment (DeeDeeG) [inline microservices/download/utils.js:121]:
```suggestion
        let name = asset?.name;
```

_If_ you want to put optional chaining back, I might have been misinformed about just how "impossible" it is for `asset` to be `undefined` here (I have found out it is indeed possible).

We're pulling data from JSON where `undefined` isn't allowed, but in general I think you had the motivation/implementation probably right before, sorry about that.

**Review by DeeDeeG — APPROVED** [review]
Approve with or without the optional chaining, feedback was addressed and implementation works in my testing, should un-break the download microservice in production.

Another good PR, lots of these flying fast around here, kudos for all the hard work.

**Comment by confused-Techie** [comment]
Thanks a ton @DeeDeeG, appreciate you and all your work on this PR as well.

Gonna go ahead and merge, then push to production!
